### PR TITLE
Optimize contextgraph batch send algorithm

### PR DESCRIPTION
**Pull request by howardjohn**
This function attempts to split the request size to be the largest
possible request that is smaller than the request size limit. It did
this by linearly checking size(list[:n]), size(list[:n-1])... This ends
up being extremely slow, because proto.size is not a very cheap
operation for such large objects. In the case of racetests, this test
was taking over 5 minutes sometimes.

This modifies the algorithm to do a binary search for the optimal
request; behavior should be the same.

With this change:
BenchmarkSendBatch/Size2-8                500000              3042 ns/op
BenchmarkSendBatch/Size200-8               10000            135250 ns/op
BenchmarkSendBatch/Size20000-8                30          52180768 ns/op

Without this change
BenchmarkSendBatch/Size2-8                500000              3082 ns/op
BenchmarkSendBatch/Size200-8               10000            140837 ns/op
BenchmarkSendBatch/Size20000-8                 1        7999998085 ns/op

This represents a 150x improvement on large request sizes and a
negligible change for small requests

Alternatively, if we don't care about the performance of this function,
we can disable it in the racetests and leave this code as is.

[x] Policies and Telemetry

### CODE REVIEW

**Comment by bianpengyuan** [comment]
Would it be nicer to calculate size of each `asst` and accumulate at append time, then create a new empty req if the current one exceed size limit? With this, req will already be well split at sending time so no search is needed.

**Comment by howardjohn** [comment]
> Would it be nicer to calculate size of each asst and accumulate at append time, then create a new empty req if the current one exceed size limit? With this, req will already be well split at sending time so no search is needed.

The thing I wasn't sure about is if the sum of the size of each `asst` is the same as the size of the full list of asst. I suppose since we have a buffer on the max size (550kb is the max in the file but a comment claims 600kb is the real max) then it could work

**Comment by bianpengyuan** [comment]
Yeah it is a rough estimate and with some head room reserved on size limit it should be safe.

File: mixer/adapter/stackdriver/contextgraph/send.go [hunk 1]
```diff
@@ -90,6 +90,27 @@ func (e edge) ToProto() *contextgraphpb.Relationship {
 	}
 }
 
+// splitBySize helps find the largest BatchRequest that is smaller than the max request size
+func splitBySizeEntity(msg *contextgraphpb.AssertBatchRequest) int {
+	curSize := proto.Size(msg)
```
Comment (bianpengyuan) [inline mixer/adapter/stackdriver/contextgraph/send.go]: This still seems to count size of each assertion for multiple times. How about this:
```
curIndex := 0
curSize := 0
for ; curIndex < len(msg.EntityPresentAssertions); curIndex++ {
  curSize += proto.Size(msg.EntityPresentAssertions[curIndex])
  if curSize >= maxReq {
    break
  }
}
return curIndex
```

Comment (howardjohn) [reply mixer/adapter/stackdriver/contextgraph/send.go]: Yeah that should be a little quicker, updated.

File: mixer/adapter/stackdriver/contextgraph/send.go [hunk 2]
```diff
@@ -90,6 +90,33 @@ func (e edge) ToProto() *contextgraphpb.Relationship {
...
 }
 
+// splitBySize helps find the largest BatchRequest that is smaller than the max request size
+func splitBySizeEntity(msg *contextgraphpb.AssertBatchRequest) int {
+	curSize := 0
+	curIndex := 0
+	for ; curIndex < len(msg.EntityPresentAssertions); curIndex++ {
+		curSize += proto.Size(msg.EntityPresentAssertions[curIndex])
+		if curSize >= maxReq {
+			curIndex--
```
Comment (bianpengyuan) [inline mixer/adapter/stackdriver/contextgraph/send.go]: It is not necessary to do `curIndex--` right? `allRequests[:split]` won't include element at index `split`.

**Pull request merged**